pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/SegmentDeletionManager.java

```diff
@@ -117,20 +117,31 @@ protected synchronized void deleteSegmentFromPropertyStoreAndLocal(String tableN
     }
 
     int nSegmentsDeleted = 0;
-    if (segmentsToDelete.size() > 0) {
+    if (!segmentsToDelete.isEmpty()) {
+      List<String> propStorePathList = new ArrayList<>(segmentsToDelete.size());
       for (String segmentId : segmentsToDelete) {
         String segmentPropertyStorePath = ZKMetadataProvider.constructPropertyStorePathForSegment(tableName, segmentId);
-        LOGGER.info("Trying to delete segment : {} from Property store.", segmentId);
-        boolean deletionFromPropertyStoreSuccessful = true;
-        if (_propertyStore.exists(segmentPropertyStorePath, AccessOption.PERSISTENT)) {
-          deletionFromPropertyStoreSuccessful = _propertyStore.remove(segmentPropertyStorePath, AccessOption.PERSISTENT);
+        propStorePathList.add(segmentPropertyStorePath);
+      }
+
+      boolean[] status = _propertyStore.remove(propStorePathList, AccessOption.PERSISTENT);
+      int nPropstoreDeleteFailed = 0;
+      for (int i = 0; i < status.length; i++) {
+        if (!status[i]) {
+          // remove API can fail because the prop store entry did not exist, so check first.
+          if (_propertyStore.exists(propStorePathList.get(i), AccessOption.PERSISTENT)) {
+            nPropstoreDeleteFailed++;
+            LOGGER.info("Could not delete {} from propertystore", propStorePathList.get(i));
+            segmentsToRetryLater.add(segmentsToDelete.get(i));
+          }
         }
+      }
 
-        if (deletionFromPropertyStoreSuccessful) {
+      for (String segmentId : segmentsToDelete) {
+        if (segmentsToRetryLater.indexOf(segmentId) < 0) {
```

Comment:
This is O(n^2), do you see this being an issue if we delete a large table?

Bonus points for coding a better algorithm on a whiteboard in less than 30 minutes :)